File: Languages/javascript/algorithms-dsa/patterns/heaps.mjs
/**
 * Heaps Pattern - Min/Max Heap implementations and common problems
 *
 * Key Concepts:
 * - A heap is a complete binary tree stored in an array
 * - MinHeap: parent <= children (root is smallest)
 * - MaxHeap: parent >= children (root is largest)
 * - Insert/Extract: O(log n), Peek: O(1)
 * - Use cases: priority queues, top-K, scheduling, medians
 */

// ============================================================
// MinHeap Class Implementation
// ============================================================

/**
 * MinHeap - smallest element always at the top.
 * Stored as array where for index i:
 *   parent = Math.floor((i - 1) / 2)
 *   left child = 2i + 1
 *   right child = 2i + 2
 */
export class MinHeap {
  constructor() {
    this.heap = [];
  }

  size() {
    return this.heap.length;
  }

  peek() {
    return this.heap[0] ?? null;
  }

  insert(val) {
    this.heap.push(val);
    this._bubbleUp(this.heap.length - 1);
  }

  extract() {
    if (this.heap.length === 0) return null;
    const min = this.heap[0];
    const last = this.heap.pop();
    if (this.heap.length > 0) {
      this.heap[0] = last;
      this._sinkDown(0);
    }
    return min;
  }

  _bubbleUp(idx) {
    while (idx > 0) {
      const parent = Math.floor((idx - 1) / 2);
      if (this.heap[parent] <= this.heap[idx]) break;
      [this.heap[parent], this.heap[idx]] = [this.heap[idx], this.heap[parent]];
      idx = parent;
    }
  }

  _sinkDown(idx) {
    const length = this.heap.length;
    while (true) {
      let smallest = idx;
      const left = 2 * idx + 1;
      const right = 2 * idx + 2;

      if (left < length && this.heap[left] < this.heap[smallest]) {
        smallest = left;
      }
      if (right < length && this.heap[right] < this.heap[smallest]) {
        smallest = right;
      }
      if (smallest === idx) break;
      [this.heap[smallest], this.heap[idx]] = [this.heap[idx], this.heap[smallest]];
      idx = smallest;
    }
  }
}

// ============================================================
// MaxHeap Class (wraps MinHeap with negation)
// ============================================================

export class MaxHeap {
  constructor() {
    this.minHeap = new MinHeap();
  }

  size() {
    return this.minHeap.size();
  }

  peek() {
    const val = this.minHeap.peek();
    return val === null ? null : -val;
  }

  insert(val) {
    this.minHeap.insert(-val);
  }

  extract() {
    const val = this.minHeap.extract();
    return val === null ? null : -val;
  }
}

// ============================================================
// Top-K Largest Elements
// ============================================================

/**
 * Find the K largest elements in an array.
 * Strategy: maintain a MinHeap of size K. If heap grows beyond K,
 * remove the smallest what remains are the K largest.
 *
 * Time: O(n log k), Space: O(k)
 */
export function topKLargest(nums, k) {
  const heap = new MinHeap();
  for (const num of nums) {
    heap.insert(num);
    if (heap.size() > k) {
      heap.extract();
    }
  }
  const result = [];
  while (heap.size() > 0) {
    result.push(heap.extract());
  }
  return result.sort((a, b) => b - a);
}

// ============================================================
// Kth Largest Element
// ============================================================

/**
 * Find the kth largest element in an unsorted array.
 * Uses a MinHeap of size K the root is the kth largest.
 *
 * Time: O(n log k), Space: O(k)
 */
export function kthLargest(nums, k) {
  const heap = new MinHeap();
  for (const num of nums) {
    heap.insert(num);
    if (heap.size() > k) {
      heap.extract();
    }
  }
  return heap.peek();
}

// ============================================================
// Merge K Sorted Arrays
// ============================================================

/**
 * Merge K sorted arrays into one sorted array.
 * Use a MinHeap storing {val, arrayIdx, elementIdx}.
 *
 * Time: O(N log k) where N = total elements, k = number of arrays
 * Space: O(k) for the heap
 */
export function mergeKSortedArrays(arrays) {
  // Custom MinHeap that compares objects by .val
  const heap = [];

  const swap = (i, j) => { [heap[i], heap[j]] = [heap[j], heap[i]]; };
  const parent = (i) => Math.floor((i - 1) / 2);
  const left = (i) => 2 * i + 1;
  const right = (i) => 2 * i + 2;

  const bubbleUp = (idx) => {
    while (idx > 0 && heap[parent(idx)].val > heap[idx].val) {
      swap(parent(idx), idx);
      idx = parent(idx);
    }
  };

  const sinkDown = (idx) => {
    const n = heap.length;
    while (true) {
      let smallest = idx;
      const l = left(idx), r = right(idx);
      if (l < n && heap[l].val < heap[smallest].val) smallest = l;
      if (r < n && heap[r].val < heap[smallest].val) smallest = r;
      if (smallest === idx) break;
      swap(smallest, idx);
      idx = smallest;
    }
  };

  const insert = (item) => { heap.push(item); bubbleUp(heap.length - 1); };
  const extract = () => {
    const min = heap[0];
    const last = heap.pop();
    if (heap.length > 0) { heap[0] = last; sinkDown(0); }
    return min;
  };

  // Initialize: insert first element from each array
  for (let i = 0; i < arrays.length; i++) {
    if (arrays[i].length > 0) {
      insert({ val: arrays[i][0], arrayIdx: i, elementIdx: 0 });
    }
  }

  const result = [];
  while (heap.length > 0) {
    const { val, arrayIdx, elementIdx } = extract();
    result.push(val);
    const nextIdx = elementIdx + 1;
    if (nextIdx < arrays[arrayIdx].length) {
      insert({ val: arrays[arrayIdx][nextIdx], arrayIdx, elementIdx: nextIdx });
    }
  }
  return result;
}

// ============================================================
// Running Median (MedianFinder)
// ============================================================

/**
 * MedianFinder - maintains running median using two heaps:
 * - maxHeap: stores the smaller half (left side)
 * - minHeap: stores the larger half (right side)
 *
 * Invariant: maxHeap.size() >= minHeap.size() (differ by at most 1)
 *
 * addNum: O(log n), findMedian: O(1)
 */
export class MedianFinder {
  constructor() {
    this.lo = new MaxHeap(); // max-heap for lower half
    this.hi = new MinHeap(); // min-heap for upper half
  }

  addNum(num) {
    // Always add to max-heap first
    this.lo.insert(num);

    // Ensure every element in lo <= every element in hi
    if (this.hi.size() > 0 && this.lo.peek() > this.hi.peek()) {
      this.hi.insert(this.lo.extract());
    } else {
      // Balance: move from lo to hi if lo is too large
      if (this.lo.size() > this.hi.size() + 1) {
        this.hi.insert(this.lo.extract());
      }
    }

    // Rebalance sizes: lo can have at most 1 more than hi
    if (this.lo.size() > this.hi.size() + 1) {
      this.hi.insert(this.lo.extract());
    } else if (this.hi.size() > this.lo.size()) {
      this.lo.insert(this.hi.extract());
    }
  }

  findMedian() {
    if (this.lo.size() > this.hi.size()) {
      return this.lo.peek();
    }
    return (this.lo.peek() + this.hi.peek()) / 2;
  }
}

// ============================================================
// Tests
// ============================================================

import { test } from "node:test";
import assert from "node:assert/strict";

test("MinHeap - insert and extract", () => {
  const h = new MinHeap();
  h.insert(5);
  h.insert(3);
  h.insert(8);
  h.insert(1);
  assert.strictEqual(h.peek(), 1);
  assert.strictEqual(h.extract(), 1);
  assert.strictEqual(h.extract(), 3);
  assert.strictEqual(h.extract(), 5);
  assert.strictEqual(h.extract(), 8);
  assert.strictEqual(h.extract(), null);
});

test("MaxHeap - insert and extract", () => {
  const h = new MaxHeap();
  h.insert(5);
  h.insert(3);
  h.insert(8);
  h.insert(1);
  assert.strictEqual(h.peek(), 8);
  assert.strictEqual(h.extract(), 8);
  assert.strictEqual(h.extract(), 5);
});

test("topKLargest - returns K largest in descending order", () => {
  assert.deepStrictEqual(topKLargest([3, 1, 5, 12, 2, 11], 3), [12, 11, 5]);
});

test("kthLargest - finds kth largest element", () => {
  assert.strictEqual(kthLargest([3, 2, 1, 5, 6, 4], 2), 5);
  assert.strictEqual(kthLargest([3, 2, 3, 1, 2, 4, 5, 5, 6], 4), 4);
});

test("mergeKSortedArrays - merges multiple sorted arrays", () => {
  assert.deepStrictEqual(
    mergeKSortedArrays([[1, 4, 7], [2, 5, 8], [3, 6, 9]]),
    [1, 2, 3, 4, 5, 6, 7, 8, 9]
  );
  assert.deepStrictEqual(
    mergeKSortedArrays([[1, 3, 5], [2, 4, 6], [0, 7]]),
    [0, 1, 2, 3, 4, 5, 6, 7]
  );
});

test("MedianFinder - running median", () => {
  const mf = new MedianFinder();
  mf.addNum(1);
  assert.strictEqual(mf.findMedian(), 1);
  mf.addNum(2);
  assert.strictEqual(mf.findMedian(), 1.5);
  mf.addNum(3);
  assert.strictEqual(mf.findMedian(), 2);
  mf.addNum(4);
  assert.strictEqual(mf.findMedian(), 2.5);
  mf.addNum(5);
  assert.strictEqual(mf.findMedian(), 3);
});

File: Languages/go/algorithms-dsa/patterns/heaps.go
package main

import (
	"container/heap"
	"fmt"
)

/*
  Pattern: Heaps (Priority Queues)
  Templates: Top-K Elements, Merge K Sorted, Running Median, Task Scheduler
*/

// MinHeap implementation for container/heap interface
type MinHeap []int

func (h MinHeap) Len() int           { return len(h) }
func (h MinHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h MinHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *MinHeap) Push(x interface{}) {
	*h = append(*h, x.(int))
}
func (h *MinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

// MaxHeap implementation
type MaxHeap []int

func (h MaxHeap) Len() int           { return len(h) }
func (h MaxHeap) Less(i, j int) bool { return h[i] > h[j] }
func (h MaxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *MaxHeap) Push(x interface{}) {
	*h = append(*h, x.(int))
}
func (h *MaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

// topKLargest - use a min-heap of size K.
// Keep only the K largest elements; the top of min-heap is the Kth largest.
func topKLargest(nums []int, k int) []int {
	h := &MinHeap{}
	heap.Init(h)
	for _, num := range nums {
		heap.Push(h, num)
		if h.Len() > k {
			heap.Pop(h)
		}
	}
	result := make([]int, h.Len())
	for i := h.Len() - 1; i >= 0; i-- {
		result[i] = heap.Pop(h).(int)
	}
	return result
}

// kthLargest - returns the kth largest element
func kthLargest(nums []int, k int) int {
	h := &MinHeap{}
	heap.Init(h)
	for _, num := range nums {
		heap.Push(h, num)
		if h.Len() > k {
			heap.Pop(h)
		}
	}
	return (*h)[0]
}

// RunningMedian - uses two heaps: maxHeap for lower half, minHeap for upper half
type MedianFinder struct {
	low  *MaxHeap // max-heap: stores smaller half
	high *MinHeap // min-heap: stores larger half
}

func NewMedianFinder() *MedianFinder {
	low := &MaxHeap{}
	high := &MinHeap{}
	heap.Init(low)
	heap.Init(high)
	return &MedianFinder{low: low, high: high}
}

func (mf *MedianFinder) AddNum(num int) {
	heap.Push(mf.low, num)
	// Balance: move max of low to high
	heap.Push(mf.high, heap.Pop(mf.low))
	// Ensure low has >= elements than high
	if mf.high.Len() > mf.low.Len() {
		heap.Push(mf.low, heap.Pop(mf.high))
	}
}

func (mf *MedianFinder) FindMedian() float64 {
	if mf.low.Len() > mf.high.Len() {
		return float64((*mf.low)[0])
	}
	return float64((*mf.low)[0]+(*mf.high)[0]) / 2.0
}

// mergeKSorted - merge k sorted arrays using a min-heap
type HeapItem struct {
	val      int
	arrayIdx int
	elemIdx  int
}
type ItemHeap []HeapItem

func (h ItemHeap) Len() int           { return len(h) }
func (h ItemHeap) Less(i, j int) bool { return h[i].val < h[j].val }
func (h ItemHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *ItemHeap) Push(x interface{}) {
	*h = append(*h, x.(HeapItem))
}
func (h *ItemHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

func mergeKSorted(arrays [][]int) []int {
	h := &ItemHeap{}
	heap.Init(h)
	for i, arr := range arrays {
		if len(arr) > 0 {
			heap.Push(h, HeapItem{val: arr[0], arrayIdx: i, elemIdx: 0})
		}
	}
	var result []int
	for h.Len() > 0 {
		item := heap.Pop(h).(HeapItem)
		result = append(result, item.val)
		if item.elemIdx+1 < len(arrays[item.arrayIdx]) {
			next := item.elemIdx + 1
			heap.Push(h, HeapItem{
				val: arrays[item.arrayIdx][next], arrayIdx: item.arrayIdx, elemIdx: next,
			})
		}
	}
	return result
}

func main() {
	// Test top-K largest
	nums := []int{3, 2, 1, 5, 6, 4}
	topK := topKLargest(nums, 3)
	fmt.Println("Top-3 largest:", topK)
	// Result contains {4, 5, 6} in sorted ascending order from min-heap pops
	topKSet := make(map[int]bool)
	for _, v := range topK {
		topKSet[v] = true
	}
	if len(topK) == 3 && topKSet[4] && topKSet[5] && topKSet[6] {
		fmt.Println("PASS: topKLargest")
	} else {
		panic(fmt.Sprintf("FAIL: topKLargest, got %v", topK))
	}

	// Test kth largest
	kth := kthLargest(nums, 2)
	if kth == 5 {
		fmt.Println("PASS: kthLargest =", kth)
	} else {
		panic(fmt.Sprintf("FAIL: kthLargest expected 5, got %d", kth))
	}

	// Test running median
	mf := NewMedianFinder()
	mf.AddNum(1)
	mf.AddNum(2)
	if mf.FindMedian() != 1.5 {
		panic("FAIL: median should be 1.5")
	}
	mf.AddNum(3)
	if mf.FindMedian() != 2.0 {
		panic("FAIL: median should be 2.0")
	}
	mf.AddNum(4)
	if mf.FindMedian() != 2.5 {
		panic("FAIL: median should be 2.5")
	}
	fmt.Println("PASS: RunningMedian")

	// Test merge K sorted arrays
	arrays := [][]int{
		{1, 4, 7},
		{2, 5, 8},
		{3, 6, 9},
	}
	merged := mergeKSorted(arrays)
	fmt.Println("Merged:", merged)
	for i := 1; i < len(merged); i++ {
		if merged[i] < merged[i-1] {
			panic("FAIL: merged not sorted")
		}
	}
	if len(merged) != 9 {
		panic("FAIL: merged length")
	}
	fmt.Println("PASS: mergeKSorted")
}
